File: java-way/src/main/java/com/crack/snap/make/easy/array/DuplicateFinder.java
package com.crack.snap.make.easy.array;

import java.util.Arrays;
import java.util.HashSet;

/**
 * @author Mohan Sharma
 */
public class DuplicateFinder {

    public boolean containsDuplicateOn2(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        for (int i=0; i < nums.length; i++) {
            for (int j = i+1; j < nums.length; j++) {
                if (nums[i] ==  nums[j])
                    return true;
            }
        }
        return false;
    }

    public boolean containsDuplicateHashing(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        var set = new HashSet<>();
        for (int num: nums) {
            if (!set.add(num)) {
                return true;
            }
        }
        return false;
    }

    public boolean containsDuplicateSort(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        for (int i = 1; i < nums.length; i++) {
            int j = i - 1;
            int key = nums[i];
            while (j >= 0 && nums[j] > key) {
                nums[j + 1] = nums [j];
                j--;
            }
            nums[j + 1] = key;
        }
        for (int i = 1; i < nums.length; i++) {
            if (nums[i-1] ==  nums[i]) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        DuplicateFinder duplicateFinder = new DuplicateFinder();
        System.out.println(duplicateFinder.containsDuplicateSort(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}));
        System.out.println(duplicateFinder.containsDuplicateSort(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 1}));
    }
}

File: python-way/com/crack/snap/make/easy/array/duplicate_finder.py
# @author Mohan Sharma

from typing import List


def contains_duplicate_on2(nums: List[int]) -> bool:
	for i in range(len(nums)):
		for j in range(i+1, len(nums)):
			if nums[i] == nums[j]:
				return True
	return False

def contains_duplicate_hashing(nums: List[int]) -> bool:
	return len(nums) != len(set(nums))

def contains_duplicate_sorting_n2(nums: List[int]) -> bool:
	# inplace sort in worst n^2
	for i in range(1, len(nums)):
		j = i-1
		key = nums[i]
		while j >=0 and nums[j] > key:
			nums[j+1] = nums[j]
			j -= 1
		nums[j + 1] = key
	
	for i in range(1, len(nums)):
		if nums[i] == nums[i-1]:
			return True
	return False

def contains_duplicate_sorting_nlogn(nums: List[int]) -> bool:
	nums.sort()
	for i in range(1, len(nums)):
		if nums[i] == nums[i-1]:
			return True
	return False

if __name__ == '__main__':
	arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
	print(contains_duplicate_sorting_nlogn(arr))
	arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1]
	print(contains_duplicate_sorting_nlogn(arr))
	arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1, 2]
	print(contains_duplicate_sorting_nlogn(arr))
